inject.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package yarpc
+
+import (
+	"fmt"
+	"reflect"
+
+	"go.uber.org/yarpc/transport"
+)
+
+var (
+	_clientFactories = make(map[reflect.Type]reflect.Value)
+	_typeOfChannel   = reflect.TypeOf((*transport.Channel)(nil)).Elem()
+)
+
+func getFactoryType(f interface{}) reflect.Type {
+	if f == nil {
+		panic("f must not be nil")
+	}
+
+	fT := reflect.TypeOf(f)
+	if fT.Kind() != reflect.Func {
+		panic(fmt.Sprintf("f must be a function, not %T", f))
+	}
+
+	if fT.NumIn() != 1 || fT.In(0) != _typeOfChannel {
+		panic(fmt.Sprintf("%v must accept only a transport.Channel", fT))
+	}
+
+	if fT.NumOut() != 1 || fT.Out(0).Kind() != reflect.Interface {
+		panic(fmt.Sprintf("%v must return a single interface result", fT))
+	}
+
+	return fT.Out(0)
+}
+
+// RegisterClientFactory registers a factory function for a specific client
+// type.
+//
+// Functions must have the signature,
+//
+// 	func(transport.Channel) T
+//
+// Where T is the type of the client. T MUST be an interface.
+//
+// This function panics if a client for the given type has already been
+// registered.
+//
+// After a factory function for a client type is registered, these objects can
+// be instantiated automatically using InjectClients.
+//
+// A function to unregister the factory function is returned. Note that the
+// function will clear whatever the corresponding type's factory function is
+// at the time it is called, regardless of whether the value matches what was
+// passed to this function or not.
+func RegisterClientFactory(f interface{}) (forget func()) {
```

Comment:
Let's change the terminology to builder in all-cases in this diff - no more factories! :P